wallet monitor

coryschwartz opened the pull request:
monitor wallet balance

Prometheus endpoint looks something like this:

```
curl localhost:8888/metrics
# HELP walletmonitor_wallet_balance wallet blance
# TYPE walletmonitor_wallet_balance gauge
walletmonitor_wallet_balance{wallet="f1xxyyzz"}  23
walletmonitor_wallet_balance{wallet="f3aabbcc"} 0
```

EDIT:

Switched from the WalletBalance API to StateGetActor so the nonce could be reported. This will make it easier to measure the message send rate as well as the balance.

The prometheus output now looks like this:
```
# HELP walletmonitor_actor_balance actor blance
# TYPE walletmonitor_actor_balance gauge
walletmonitor_actor_balance{address="f1xyzabc"} 16
# HELP walletmonitor_actor_nonce nonce
# TYPE walletmonitor_actor_nonce gauge
walletmonitor_actor_nonce{address="f1xyzabc"} 211

```

magik6k commented on Makefile:
```diff
@@ -221,6 +221,11 @@ testground:
+wallet-monitor:
+	go build -o wallet-monitor ./cmd/wallet-monitor
+.PHONY: wallet-monitor
+BINS+=wallet-monitor
```
I wonder if it would maybe be better to call this `lotus-monitor`, which would let us expand this as a general utility.

I'd really love to be able to just run `lotus-monitor config.toml` on my miner, with the config containing node API infos, and what to monitor on each node - wallet balances, miner power, deal counts, sector states, etc, etc.

Stebalien commented:
Converting to draft while feedback is addressed.

coryschwartz commented:
I've added the ability to monitor for more than just wallet addresses and datacap monitoring. Want to run through a few more tests and then will take it out of draft.

coryschwartz commented:
This is ready for review now.

This will be used to export wallet fil and datacap so they can be monitored and alerted on. I added a couple of miner metrics as well, but I don't have a specific application for it in mind, maybe there are different or better metrics that can be added now or in the future.

This will be immediely usable by PL projects and I'll add it to our generic "lotus-bundle" helm chart to monitor the wallets that are created with those deployments.

There is a gotcha about using the gateway. Its limited API lacks some endpoints that might be interesting to someone running a miner. I added a `--gateway` to avoid using endpoints that aren't  on the gateway. Commmand line flags can be provided as env vars to make it easy to start with a systemd unit in a docker container.

This example would monitor the top 3 miners with the most power and top 3 on the "rich list" on https://filfox.info 

```
FULLNODE_API_INFO=ws://127.0.0.1:1234 GOLOG_LOG_LEVEL=debug  ./lotus-monitor --listen :8888 --actor f0121 -
-actor f0118 --actor f0119 --miner f0127595 --miner f0123261 --miner f0142720
```

and the prometheus output looks as follows

```
curl localhost:8888/metrics
# HELP lotusmonitor_actor_Balance actor blance
# TYPE lotusmonitor_actor_Balance gauge
lotusmonitor_actor_Balance{actorAddress="f0118"} 4.6629006e+07
lotusmonitor_actor_Balance{actorAddress="f0119"} 4.4266986e+07
lotusmonitor_actor_Balance{actorAddress="f0121"} 9.8048732e+07
# HELP lotusmonitor_actor_Datacap client datacap
# TYPE lotusmonitor_actor_Datacap gauge
lotusmonitor_actor_Datacap{actorAddress="f0118"} 0
lotusmonitor_actor_Datacap{actorAddress="f0119"} 0
lotusmonitor_actor_Datacap{actorAddress="f0121"} 0
# HELP lotusmonitor_actor_Nonce nonce
# TYPE lotusmonitor_actor_Nonce gauge
lotusmonitor_actor_Nonce{actorAddress="f0118"} 0
lotusmonitor_actor_Nonce{actorAddress="f0119"} 0
lotusmonitor_actor_Nonce{actorAddress="f0121"} 0
# HELP lotusmonitor_actor_QualityAdjPower raw power
# TYPE lotusmonitor_actor_QualityAdjPower gauge
lotusmonitor_actor_QualityAdjPower{actorAddress="f0118"} 0
lotusmonitor_actor_QualityAdjPower{actorAddress="f0119"} 0
lotusmonitor_actor_QualityAdjPower{actorAddress="f0121"} 0
# HELP lotusmonitor_actor_RawPower raw power
# TYPE lotusmonitor_actor_RawPower gauge
lotusmonitor_actor_RawPower{actorAddress="f0118"} 0
lotusmonitor_actor_RawPower{actorAddress="f0119"} 0
lotusmonitor_actor_RawPower{actorAddress="f0121"} 0
# HELP lotusmonitor_miner_Balance miner fil balance
# TYPE lotusmonitor_miner_Balance gauge
lotusmonitor_miner_Balance{minerAddress="f0123261"} 34001.958628288805
lotusmonitor_miner_Balance{minerAddress="f0127595"} 16233.416964258538
lotusmonitor_miner_Balance{minerAddress="f0142720"} 2520.4754734352036
# HELP lotusmonitor_miner_Faults faults
# TYPE lotusmonitor_miner_Faults gauge
lotusmonitor_miner_Faults{minerAddress="f0123261"} 1
lotusmonitor_miner_Faults{minerAddress="f0127595"} 0
lotusmonitor_miner_Faults{minerAddress="f0142720"} 2
# HELP lotusmonitor_miner_QualityAdjPower quality-adjusted power
# TYPE lotusmonitor_miner_QualityAdjPower gauge
lotusmonitor_miner_QualityAdjPower{minerAddress="f0123261"} 1.2960582647729357e+17
lotusmonitor_miner_QualityAdjPower{minerAddress="f0127595"} 1.494026020548772e+17
lotusmonitor_miner_QualityAdjPower{minerAddress="f0142720"} 1.220312908735447e+17
# HELP lotusmonitor_miner_RawPower raw power
# TYPE lotusmonitor_miner_RawPower gauge
lotusmonitor_miner_RawPower{minerAddress="f0123261"} 1.2960582647729357e+17
lotusmonitor_miner_RawPower{minerAddress="f0127595"} 1.494026020548772e+17
lotusmonitor_miner_RawPower{minerAddress="f0142720"} 1.220312908735447e+17
# HELP lotusmonitor_miner_Recoveries recoveries
# TYPE lotusmonitor_miner_Recoveries gauge
lotusmonitor_miner_Recoveries{minerAddress="f0123261"} 0
lotusmonitor_miner_Recoveries{minerAddress="f0127595"} 0
lotusmonitor_miner_Recoveries{minerAddress="f0142720"} 1
```

coryschwartz commented:
@filecoin-project/lotus-maintainers can I have a re-review? Just made the linter pass.

coryschwartz commented on cmd/lotus-monitor/util.go:
```diff
@@ -0,0 +1,63 @@
+	return types.BigDivFloat(attofil, types.FromFil(1))
+}
+
+func verifiedPower(ctx context.Context, api v0api.FullNode, addr address.Address) (bool, abi.StoragePower, error) {
```
Yes it should; I don't know what I was thinking.

coryschwartz commented on cmd/lotus-monitor/actors.go:
```diff
@@ -0,0 +1,111 @@
+	}
+	if actor == nil {
+		log.Warnw("actor not found", "actor", addr)
+		errs <- err
```
Thanks.

coryschwartz commented on cmd/lotus-monitor/actors.go:
```diff
@@ -0,0 +1,111 @@
+		errs <- err
+	} else {
+		stats.Record(ctx, actorQualityAdjPowerMetric.M(power.Int64()))
+		stats.Record(ctx, actorRawPowerMetric.M(power.Int64()))
```
you're right. Thanks.

rjan90 commented:
Closing this PR as stale/outdated
